### PR TITLE
Refs #12556 - remove System from views & controllers

### PR DESCRIPTION
**Pull request by jlsherrill**
This attempts to remove any references to the System
model from all controllers and views.  The only exception to
this is the CandlepinProxies controller which relies on creation and
updating of the models during sub-man actions.  These will be removed
along with the model.

Some areas to test more heavily are:
  * Content View Version Removal
  * Various Content Host pages

A followup commit will remove the model and tests.

### CODE REVIEW

File: app/models/katello/concerns/content_facet_host_extensions.rb [hunk 1]
```diff
@@ -53,6 +53,13 @@ def find_by_installable_errata(_key, operator, value)
             { :conditions => "#{::Host::Managed.table_name}.id IN (#{facets.pluck(:host_id).join(',')})" }
           end
         end
+
+        def in_content_view_environment(content_view = nil, lifecycle_environment = nil)
+          relation = self.joins(:content_facet)
+          relation = relation.where("#{::Katello::Host::ContentFacet.table_name}.content_view_id" => content_view) if content_view
+          relation = relation.where("#{::Katello::Host::ContentFacet.table_name}.lifecycle_environment_id" => lifecycle_environment) if lifecycle_environment
+          relation
+        end
```
Comment (jlsherrill) [inline app/models/katello/concerns/content_facet_host_extensions.rb]: TODO: write a test

File: app/models/katello/content_view_version.rb [hunk 2]
```diff
@@ -280,6 +280,11 @@ def validate_destroyable!(skip_environment_check = false)
       true
     end
 
+    def all_hosts_editable?(lifecycle_environment)
+      ::Host.in_content_view_environment(self.content_view, lifecycle_environment).count ==
+          ::Host.authorized("view_hosts").in_content_view_environment(self.content_view, lifecycle_environment).count
+    end
```
Comment (jlsherrill) [inline app/models/katello/content_view_version.rb]: TODO: write a test

File: app/models/katello/glue/candlepin/pool.rb [hunk 3]
```diff
@@ -117,8 +117,10 @@ def import_data
         self.create_activation_key_associations
       end
 
-      def systems
-        System.all_by_pool(cp_id)
+      def hosts
+        entitlements = Resources::Candlepin::Entitlement.get
+        uuids = entitlements.delete_if { |ent| ent["pool"]["id"] != cp_id }.map { |ent| ent["consumer"]["uuid"] }
+        ::Host.where(:id => Katello::Host::ContentFacet.where(:uuid => uuids).pluck(:host_id))
       end
```
Comment (jlsherrill) [inline app/models/katello/glue/candlepin/pool.rb]: TODO write a test

File: app/controllers/katello/api/v2/products_controller.rb [hunk 4]
```diff
@@ -135,14 +133,6 @@ def find_activation_key
       end
     end
 
-    def find_system
```
Comment (parthaa) [inline app/controllers/katello/api/v2/products_controller.rb]: find_host not needed eh ?

Comment (jlsherrill) [inline app/controllers/katello/api/v2/products_controller.rb]: didn't seem to be used anywhere.

File: app/controllers/katello/concerns/authorization/api/v2/content_views_controller.rb [hunk 5]
```diff
@@ -118,14 +118,17 @@ def authorize_remove_environments(view, options)
       # and also ensure that the environments have the remove permission
       return deny_access unless KTEnvironment.promotable.where(:id => env_ids).count == env_ids.size && view.promotable_or_removable?
 
-      if Katello::System.where(:content_view_id => view, :environment_id => env_ids).count > 0
+      total_count = Katello::Host::ContentFacet.where(:content_view_id => view, :lifecycle_environment_id => env_ids).count
+      if total_count > 0
         unless options[:system_content_view_id] && options[:system_environment_id]
```
Comment (parthaa) [inline app/controllers/katello/concerns/authorization/api/v2/content_views_controller.rb]: Should these be host_content_view_id and host_environment_id

Comment (jlsherrill) [inline app/controllers/katello/concerns/authorization/api/v2/content_views_controller.rb]: i debated changing this, but given the scale of the change I was hesitant to do so.

Comment (jlsherrill) [inline app/controllers/katello/concerns/authorization/api/v2/content_views_controller.rb]: and the api breakage it would result in, open to suggestions/thoughts though

File: app/models/katello/content_view_version.rb [hunk 6]
```diff
@@ -280,6 +280,11 @@ def validate_destroyable!(skip_environment_check = false)
       true
     end
 
+    def all_hosts_editable?(lifecycle_environment)
+      ::Host.in_content_view_environment(self.content_view, lifecycle_environment).count ==
+          ::Host.authorized("view_hosts").in_content_view_environment(self.content_view, lifecycle_environment).count
```
Comment (parthaa) [inline app/models/katello/content_view_version.rb]: Weird question but what happens if there are 0 hosts. And user is not authorized mess with any hosts. Wont this still return true.

Comment (jlsherrill) [inline app/models/katello/content_view_version.rb]: it would.  What its trying to figure out is if the user has edit access to all hosts associated with that content view and lifecycle env.  If there are 0 hosts then the user has access to edit all hosts.

File: app/models/katello/glue/candlepin/pool.rb [hunk 7]
```diff
@@ -135,12 +137,6 @@ def create_activation_key_associations
       def host
         System.find_by(:uuid => host_id) if host_id
```
Comment (parthaa) [inline app/models/katello/glue/candlepin/pool.rb]: Handling this elsewhere ?

Comment (jlsherrill) [inline app/models/katello/glue/candlepin/pool.rb]: that's a model/glue change that will be fixed in the next PR.

File: app/views/katello/api/v2/content_view_versions/base.json.rabl [hunk 8]
```diff
@@ -47,13 +47,13 @@ child :environments => :environments do
     {
       :readable => env.readable?,
       :promotable_or_removable => env.promotable_or_removable?,
-      :all_systems_editable => Katello::System.all_editable?(version.content_view_id, env.id),
-      :all_keys_editable => Katello::System.all_editable?(version.content_view_id, env.id)
```
Comment (parthaa) [inline app/views/katello/api/v2/content_view_versions/base.json.rabl]: wow

**Comment by parthaa** [comment]
LGTM code wise.. nothing stood out. Will run it on my dev setup and review further.

**Comment by jlsherrill** [comment]
Test written

File: app/controllers/katello/api/v2/systems_controller.rb [hunk 9]
```diff
@@ -1,213 +0,0 @@
...
-        @content_view = nil
-      end
-    end
-
-    def authorize_environment
-      return deny_access unless @environment.readable?
-      true
-    end
-  end
-end
```
Comment (waldenraines) [inline app/controllers/katello/api/v2/systems_controller.rb]: 👍

File: app/lib/actions/katello/content_view_environment/reassign_objects.rb [hunk 10]
```diff
@@ -4,8 +4,8 @@ module ContentViewEnvironment
       class ReassignObjects < Actions::Base
         def plan(content_view_environment, options)
           concurrence do
-            content_view_environment.systems.each do |system|
-              plan_action(System::Reassign, system, options[:system_content_view_id], options[:system_environment_id])
+            content_view_environment.hosts.each do |host|
+              plan_action(Host::Reassign, host, options[:system_content_view_id], options[:system_environment_id])
```
Comment (waldenraines) [inline app/lib/actions/katello/content_view_environment/reassign_objects.rb]: Is there even a host reassign task?  I see one for systems in but not for hosts.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/details/content-host-add-subscriptions.controller.js [hunk 11]
```diff
@@ -14,8 +14,8 @@
  *   Provides the functionality for the content host details action pane.
  */
 angular.module('Bastion.content-hosts').controller('ContentHostAddSubscriptionsController',
```
Comment (waldenraines) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/details/content-host-add-subscriptions.controller.js]: If we want to rename the controllers in the UI in a future PR I am happy to do it.

File: app/controllers/katello/concerns/authorization/api/v2/content_views_controller.rb (hunk 5 above)
Comment (ehelms) [inline app/controllers/katello/concerns/authorization/api/v2/content_views_controller.rb]: I'd be OK with a separate PR that adds a new parameter and deprecates this parameter.

**Comment by ehelms** [comment]
Still see this in my UI -- https://github.com/Katello/katello/blob/1a0abd7eec6a9a2d31ebd2b272800ef5111e199e/engines/bastion_katello/app/assets/javascripts/bastion_katello/content-views/deletion/views/version-deletion-environments.html#L30

**Comment by ehelms** [comment]
I also got this in the content view remove workflow:

```
NoMethodError: undefined method `in_content_view_environment' for Katello::Host:Module
 | /home/vagrant/katello/app/models/katello/content_view_environment.rb:38:in `hosts'
 | /home/vagrant/katello/app/lib/actions/katello/content_view_environment/reassign_objects.rb:7:in `block in plan'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/execution_plan.rb:281:in `call'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/execution_plan.rb:281:in `switch_flow'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/action.rb:364:in `concurrence'
 | /home/vagrant/katello/app/lib/actions/katello/content_view_environment/reassign_objects.rb:6:in `plan'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/action.rb:461:in `block (3 levels) in execute_plan'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/middleware/stack.rb:26:in `call'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/middleware/stack.rb:26:in `pass'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/middleware.rb:17:in `pass'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/middleware.rb:34:in `plan'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/middleware/stack.rb:22:in `call'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/middleware/stack.rb:26:in `pass'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/middleware.rb:17:in `pass'
 | /home/vagrant/katello/app/lib/actions/middleware/keep_locale.rb:7:in `plan'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/middleware/stack.rb:22:in `call'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/middleware/stack.rb:26:in `pass'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/middleware.rb:17:in `pass'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/middleware.rb:34:in `plan'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/middleware/stack.rb:22:in `call'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/middleware/world.rb:30:in `execute'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/action.rb:460:in `block (2 levels) in execute_plan'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/execution_plan.rb:281:in `call'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/execution_plan.rb:281:in `switch_flow'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/action.rb:364:in `concurrence'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/action.rb:459:in `block in execute_plan'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/action.rb:419:in `call'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/action.rb:419:in `block in with_error_handling'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/action.rb:419:in `catch'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/action.rb:419:in `with_error_handling'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/action.rb:458:in `execute_plan'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/action.rb:260:in `execute'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/execution_plan/steps/plan_step.rb:54:in `block in execute'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/execution_plan/steps/abstract.rb:155:in `call'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/execution_plan/steps/abstract.rb:155:in `with_meta_calculation'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/execution_plan/steps/plan_step.rb:53:in `execute'
 | /home/vagrant/.rvm/gems/ruby-2.2.1/gems/dynflow-0.8.11/lib/dynflow/action.rb:392:in `plan_action'
 | /home/vagrant/katello/app/lib/actions/katello/content_view/remove.rb:32:in `block (3 levels) in plan'
```

**Comment by jlsherrill** [comment]
@ehelms updated to address your two comments

File: app/controllers/katello/concerns/authorization/api/v2/content_views_controller.rb (hunk 5 above)
Comment (jlsherrill) [inline app/controllers/katello/concerns/authorization/api/v2/content_views_controller.rb]: opened http://projects.theforeman.org/issues/15650

File: app/models/katello/concerns/content_facet_host_extensions.rb [hunk 12]
```diff
@@ -53,6 +53,13 @@ def find_by_installable_errata(_key, operator, value)
             { :conditions => "#{::Host::Managed.table_name}.id IN (#{facets.pluck(:host_id).join(',')})" }
           end
         end
+
+        def in_content_view_environment(content_view = nil, lifecycle_environment = nil)
```
Comment (ehelms) [inline app/models/katello/concerns/content_facet_host_extensions.rb]: I feel like this should maybe just return the relation or at least require one of these when calling it? Seems strange to allow both to be nil when its specifically 'in_content_view_environment' and not in content view and/or environment?

Comment (jlsherrill) [inline app/models/katello/concerns/content_facet_host_extensions.rb]: If both are nil, the joins() actually restricts to any hosts that have a content_view_environment.  This may be useful, but a user of the function may not know what to expect with that input.  I'd probably rather error if both are nil.  Thoughts?

Comment (ehelms) [inline app/models/katello/concerns/content_facet_host_extensions.rb]: So:

````
Host.in_content_view_environment
```

Returns all hosts that have a content view environment? 

Does the following return all hosts with a content view environment in environment 1?

```
Host.in_content_view_environment(nil, 1)
```

Comment (jlsherrill) [inline app/models/katello/concerns/content_facet_host_extensions.rb]: Yes and Yes

Comment (ehelms) [inline app/models/katello/concerns/content_facet_host_extensions.rb]: As a method signature it's not very easy to read (I have to go look up the method) to know what I might be nil-ing the value of, bit of a nitpick I suppose but would make this easier to read in places where one or the other parameter is nil.

Comment (jlsherrill) [inline app/models/katello/concerns/content_facet_host_extensions.rb]: 'would make this easier to read'   what would?

Comment (ehelms) [inline app/models/katello/concerns/content_facet_host_extensions.rb]: Not using positional parameters that are expected to be nil:

```
Host.in_content_view_environment(nil, env)
Host.in_content_view_environment(cv)
```

versus

```
Host.in_content_view_environment(:lifecycle_environment => env)
Host.in_content_view_environment(:content_view => cv)
```

Comment (jlsherrill) [inline app/models/katello/concerns/content_facet_host_extensions.rb]: Ah yeah i thought about that, with two arguments its kinda a toss to me.  I'd still have to look at the method signature i think to make sure i was passing the right thing.  I'm good with either

Comment (ehelms) [inline app/models/katello/concerns/content_facet_host_extensions.rb]: When 1 or both can be nil, I think the number of arguments diminishes the reason to do it given you have to pass nil which isn't obvious. What I might ask for instead of changing the method signature is to use keywords where the nil is since this is Ruby 2, e.g. 

```
Host.in_content_view_environment(:lifecycle_environment => env)
```

The above should work out of the both without changing the method signature.

Comment (jlsherrill) [inline app/models/katello/concerns/content_facet_host_extensions.rb]: updated

**Comment by ehelms** [comment]
ACK for me, @parthaa @waldenraines  are you OK with this?

**Comment by parthaa** [comment]
ACK

**Comment by jlsherrill** [comment]
Merging since @waldenraines is on PTO and his comments were more supportive :)
